### ISSUE TITLE
[2.1.2 CE] env.php overwritten when a module is disable and cron runs in the background

### ISSUE THREAD

**Issue by azeemism** [magento/magento2#7165]
### Summary

env.php overwritten when a module is disable and cron runs in the background
### Preconditions
1. Composer install Magento 2.1.1 CE, 2.1.2 CE
2. php bin/magento setup:install --backend-frontname="123adcd"
3. php bin/magento deploy:mode:set developer
4. php bin/magento maintenance:enable
5. php bin/magento setup:di:compile
6. php bin/magento --ansi setup:static-content:deploy en_CA
7. php bin/magento indexer:reindex
8. php bin/magento cache:disable
9. php bin/magento cache:flush
10. mariadb

```
root@skynet:~# dpkg -l | grep mariadb
ii  libmariadbclient18                   10.1.17+maria-1~jessie               amd64        MariaDB database client library
ii  mariadb-client                       10.1.17+maria-1~jessie               all          MariaDB database client (metapackage depending on the latest version)
ii  mariadb-client-10.1                  10.1.17+maria-1~jessie               amd64        MariaDB database client binaries
ii  mariadb-client-core-10.1             10.1.17+maria-1~jessie               amd64        MariaDB database core client binaries
ii  mariadb-common                       10.1.17+maria-1~jessie               all          MariaDB database common files (e.g. /etc/mysql/conf.d/mariadb.cnf)
ii  mariadb-server                       10.1.17+maria-1~jessie               all          MariaDB database server (metapackage depending on the latest version)
ii  mariadb-server-10.1                  10.1.17+maria-1~jessie               amd64        MariaDB database server binaries
ii  mariadb-server-core-10.1             10.1.17+maria-1~jessie               amd64        MariaDB database core server files
```
1. php7.0

```
root@skynet:~# dpkg -l | grep php7
ii  libapache2-mod-php7.0                7.0.10-1~dotdeb+8.1                  amd64        server-side, HTML-embedded scripting language (Apache 2 module)
ii  php7.0                               7.0.10-1~dotdeb+8.1                  all          server-side, HTML-embedded scripting language (metapackage)
ii  php7.0-bcmath                        7.0.10-1~dotdeb+8.1                  amd64        Bcmath module for PHP
ii  php7.0-cli                           7.0.10-1~dotdeb+8.1                  amd64        command-line interpreter for the PHP scripting language
ii  php7.0-common                        7.0.10-1~dotdeb+8.1                  amd64        documentation, examples and common module for PHP
ii  php7.0-curl                          7.0.10-1~dotdeb+8.1                  amd64        CURL module for PHP
ii  php7.0-dev                           7.0.10-1~dotdeb+8.1                  amd64        Files for PHP7.0 module development
ii  php7.0-fpm                           7.0.10-1~dotdeb+8.1                  amd64        server-side, HTML-embedded scripting language (FPM-CGI binary)
ii  php7.0-gd                            7.0.10-1~dotdeb+8.1                  amd64        GD module for PHP
ii  php7.0-igbinary                      1.2.1-1~2b7c703-1~dotdeb+8.2         amd64        igbinary serializer for PHP
ii  php7.0-imagick                       3.4.2-1~dotdeb+8.2                   amd64        Provides a wrapper to the ImageMagick library
ii  php7.0-intl                          7.0.10-1~dotdeb+8.1                  amd64        Internationalisation module for PHP
ii  php7.0-json                          7.0.10-1~dotdeb+8.1                  amd64        JSON module for PHP
ii  php7.0-mbstring                      7.0.10-1~dotdeb+8.1                  amd64        MBSTRING module for PHP
ii  php7.0-mcrypt                        7.0.10-1~dotdeb+8.1                  amd64        libmcrypt module for PHP
ii  php7.0-mysql                         7.0.10-1~dotdeb+8.1                  amd64        MySQL module for PHP
ii  php7.0-opcache                       7.0.10-1~dotdeb+8.1                  amd64        Zend OpCache module for PHP
ii  php7.0-pspell                        7.0.10-1~dotdeb+8.1                  amd64        pspell module for PHP
ii  php7.0-readline                      7.0.10-1~dotdeb+8.1                  amd64        readline module for PHP
ii  php7.0-redis                         3.0.0~rc1-1~dotdeb+8.2               amd64        PHP extension for interfacing with Redis
ii  php7.0-soap                          7.0.10-1~dotdeb+8.1                  amd64        SOAP module for PHP
ii  php7.0-xml                           7.0.10-1~dotdeb+8.1                  amd64        DOM, SimpleXML, WDDX, XML, and XSL module for PHP
ii  php7.0-xmlrpc                        7.0.10-1~dotdeb+8.1                  amd64        XMLRPC-EPI module for PHP
ii  php7.0-xsl                           7.0.10-1~dotdeb+8.1                  all          XSL module for PHP (dummy)
ii  php7.0-zip                           7.0.10-1~dotdeb+8.1                  amd64        Zip module for PHP
```
1. Debian

```
root@skynet:~# cat /proc/version
Linux version 3.16.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.8.4 (Debian 4.8.4-1) ) #1 SMP Debian 3.16.7-ckt25-2+deb8u3 (2016-07-02)


root@skynet:~# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.5 (jessie)
Release:        8.5
Codename:       jessie
```
1. Composer version 1.2.1 2016-09-12 11:27:19
2. opcache enabled
3. Nginx

```
root@skynet:~# dpkg -l | grep nginx
ii  libnginx-mod-http-auth-pam           1.10.1-1~dotdeb+8.2                  amd64        PAM authentication module for Nginx
ii  libnginx-mod-http-geoip              1.10.1-1~dotdeb+8.2                  amd64        GeoIP HTTP module for Nginx
ii  libnginx-mod-http-image-filter       1.10.1-1~dotdeb+8.2                  amd64        HTTP image filter module for Nginx
ii  libnginx-mod-http-lua                1.10.1-1~dotdeb+8.2                  amd64        LUA module for Nginx
ii  libnginx-mod-http-ndk                1.10.1-1~dotdeb+8.2                  amd64        Nginx Development Kit module
ii  libnginx-mod-http-perl               1.10.1-1~dotdeb+8.2                  amd64        Perl module for Nginx
ii  libnginx-mod-http-xslt-filter        1.10.1-1~dotdeb+8.2                  amd64        XSLT Transformation module for Nginx
ii  libnginx-mod-mail                    1.10.1-1~dotdeb+8.2                  amd64        Mail module for Nginx
ii  libnginx-mod-stream                  1.10.1-1~dotdeb+8.2                  amd64        Stream module for Nginx
ii  nginx-common                         1.10.1-1~dotdeb+8.2                  all          small, powerful, scalable web/proxy server - common files
ii  nginx-extras                         1.10.1-1~dotdeb+8.2                  amd64        nginx web/proxy server (extended version)
```
### Steps to reproduce

1. Setup Cron to run in the background. 
   As "root" `nano /etc/cron.d/magento_cron`
   Add:

```
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/bin/magento cron:run | grep -v -e "Ran jobs by schedule" >> /var/www/example.com/magento/2.x/2.x/var/log/magento.cron.log  2>&1
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/update/cron.php >> /var/www/example.com/magento/2.x/2.x/var/log/update.cron.log  2>&1
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/bin/magento setup:cron:run >> /var/www/example.com/magento/2.x/2.x/var/log/setup.cron.log  2>&1
```

Restart cron:

```
systemctl restart cron
```
1. Disable an extension

```
php bin/magento module:disable Magefan_Blog
```

Note: this is a free extension.
1. Let cron run in the background once, or several times
2. run

```
php bin/magento module:status
```

Note: File permission are all setup correctly with `www-data` as the owner and the group
Note: Only tested with 3rd party extensions/modules.
### Expected result

1. Module are listed with their status
2. Cron runs in the background without any errors
3. `env.php` is not overwritten
### Actual result

1. Running `php bin/magento module:status` results in the following message:

```
  [InvalidArgumentException]
  There are no commands defined in the "module" namespace.
```
1. The cron log shows the following:

```
  [InvalidArgumentException]
  There are no commands defined in the "setup:cron" namespace.
```
1. `${webroot}/app/etc/env.php` is overwritten sometimes partially sometimes completely
### Additional comments / workaround

A manual workaround is to stop cron from running in the background when disabling a module i.e.

As `root`:

```
systemctl stop cron
```

As `www-data`:

```
php bin/magento deploy:mode:set developer
php bin/magento cache:disable
php bin/magento cache:flush
php bin/magento module:disable Magefan_Blog
php bin/magento module:status
php bin/magento setup:upgrade
php bin/magento setup:di:compile
php bin/magento --ansi setup:static-content:deploy en_CA
php bin/magento indexer:reindex
php bin/magento cache:flush
```

As `root`:

```
systemctl restart nginx php7.0-fpm
systemctl start cron
```

**Comment by BaDos** [comment]
Hi @azeemism 

Thank you for your feedback.
I tried to reproduce it with/without Magefan_Blog module, but I cannot reproduce it.

When did you install Magefan_Blog? Please write number step.
This information is absent on preconditions list.

You enabled maintenance mode `php bin/magento maintenance:enable`
Did you run `php bin/magento module:disable Magefan_Blog` and  `php bin/magento module:status` when this mode is enabled?

**Comment by azeemism** [comment]
Hi @BaDos ,

Thanks for checking on the case.

Yes I enabled maintenance mode with `php bin/magento maintenance:enable`
Yes I ran `php bin/magento module:disable Magefan_Blog` and `php bin/magento module:status` when this mode is enabled?

Steps to reproduce:
1. Use php7.0 w/ opcache
2. Install magento 2 with composer
3. `php bin/magento maintenance:enable`
4. `php bin/magento deploy:mode:set developer`
5. As root copy magefan Blog files to `app/code` cd to webroot and `find . -type d -exec chmod 770 {} \; && find . -type f -exec chmod 660 {} \; && chown -R www-data:www-data .`
6. `php bin/magento setup:upgrade`
7. `php bin/magento setup:di:compile`
8. `php bin/magento --ansi setup:static-content:deploy en_CA`
9. `php bin/magento indexer:reindex`
10. `php bin/magento cache:disable`
11. `php bin/magento cache:flush`
12. Run again to ensure you are still in maintenance mode: `php bin/magento maintenance:enable`
13. Setup cron to run every minute and mail you STDERR:

```
# Send output to email address, as long as output (STDOUT or STDERR or both are not sent to /dev/nul)
MAILTO=you@your_domain.com
#
# m h dom mon dow user command
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/bin/magento cron:run | grep -v -e "Ran jobs by schedule" >> /var/www/example.com/magento/2.x/2.x/var/log/magento.cron.log  2>&1
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/update/cron.php >> /var/www/example.com/magento/2.x/2.x/var/log/update.cron.log  2>&1
*/1 * * * * www-data /usr/bin/php /var/www/example.com/magento/2.x/2.x/bin/magento setup:cron:run >> /var/www/example.com/magento/2.x/2.x/var/log/setup.cron.log  2>&1

```
1. As root `systemctl restart cron`
2. Ensure that the module is installed `php bin/magento module:status`
3. Disable the module `php bin/magento module:disable Magefan_Blog`
4. You will get the following message:

```
The following modules have been disabled:
- Magefan_Blog

Cache cleared successfully.
Generated classes cleared successfully. Please run the 'setup:di:compile' command to generate classes.
Info: Some modules might require static view files to be cleared. To do this, run 'module:disable' with the --clear-static-content option to clear them.

```
1. Wait 5 mins
2. Now you will get flooded with emails just about every minute that look like this

```
PHP Parse error:  syntax error, unexpected '=>' (T_DOUBLE_ARROW), expecting ',' or ')' in /var/www/example.com/magento/2.x/2.x/app/etc/env.php on line 4
```
1. And you will notice that app/etc/config.php will have changed. Below is a snapshot of the file before disabling and just after disabling"

```
www-data@skynet:~/example.com/magento/2.x/2.x/app/etc$ ls -l
total 104
-rw-rw---- 1 www-data www-data  3776 Nov  4 15:07 config.php
-rw-rw---- 1 www-data www-data 77663 Oct 21 15:20 di.xml
-rw-rw---- 1 www-data www-data  2699 Nov  4 15:05 env.php
-rw-rw---- 1 www-data www-data  2699 Oct 12 12:35 env.php_bck
-rw-rw---- 1 www-data www-data  2732 Oct 12 11:59 env.php_bck_w_issue_20161012
-rw-rw---- 1 www-data www-data    16 Oct 24 09:52 env.php_bck_w_issue_20161024
-rw-rw---- 1 www-data www-data   859 Oct 21 15:20 NonComposerComponentRegistration.php
-rw-rw---- 1 www-data www-data    70 Nov  3 16:07 vendor_path.php
www-data@skynet:~/example.com/magento/2.x/2.x/app/etc$ diff env.php env.php_bck
www-data@skynet:~/example.com/magento/2.x/2.x/app/etc$ ls -l
total 104
-rw-rw---- 1 www-data www-data  3776 Nov  4 15:40 config.php
-rw-rw---- 1 www-data www-data 77663 Oct 21 15:20 di.xml
-rw-rw---- 1 www-data www-data  2699 Nov  4 15:41 env.php
-rw-rw---- 1 www-data www-data  2699 Oct 12 12:35 env.php_bck
-rw-rw---- 1 www-data www-data  2732 Oct 12 11:59 env.php_bck_w_issue_20161012
-rw-rw---- 1 www-data www-data    16 Oct 24 09:52 env.php_bck_w_issue_20161024
-rw-rw---- 1 www-data www-data   859 Oct 21 15:20 NonComposerComponentRegistration.php
-rw-rw---- 1 www-data www-data    70 Nov  3 16:07 vendor_path.php
www-data@skynet:~/example.com/magento/2.x/2.x/app/etc$ diff env.php env.php_bck
2,3c2
< return 1;
< ay (
---
> return array (
www-data@skynet:~/example.com/magento/2.x/2.x/app/etc$
```

So the original looked like:

```
<?php
return array (
  'backend' =>
  array (
    'frontName' => 'myadmin',
```

Now it looks like:

```
<?php
return 1;
ay (
  'backend' =>
  array (
    'frontName' => 'myadmin',
```
1. While this is an example of the change to the file, I have had the file content completely wiped including the 

```
  'crypt' =>
  array (
    'key' => 'somealphanumericnumber',
```

Which, would leave one completely cluster "bombed" if they did not keep a record of it.

NOTE: If the module is disabled for `config.php` set as follows `'Magefan_Blog' => 0,` then env.php does not get overwritten...and I am not sure why? What is the difference between setting `'Magefan_Blog' => 0,` in ``config.php` vs `php bin/magento module:disable Magefan_Blog`?

**Comment by shiftedreality** [comment]
Hi @azeemism 

I was unable to reproduce your issue also.
Deep investigating gave me ideas about cache enable-disable.
Could you please try `bin/magento cache:disable` and then `bin/magento cache:enable`?

**Comment by peterhirn** [comment]
After a `composer-update` on 2016-12-03 06:30 my `env.php` was replaced with:

    <?php
    return 1;

I replaced the configuration and now `setup:upgrade` is broken:

    Can't run this operation: deployment configuration is absent. Run 'magento setup:config:set --help' for options.

And the `cache` namespace is missing:

    There are no commands defined in the "cache" namespace.

**Comment by Ctucker9233** [comment]
@peterhirn Try deleting var/di then run the php bin/magento setup:upgrade command. After that, your cache commands should function again.

**Comment by peterhirn** [comment]
Turns out all modules were disabled. After running `magento module:enable --all` everything is working again.

**Comment by azeemism** [comment]
@shiftedreality 
already using `php bin/magento cache:disable` in steps to reproduce (see above). Am currently running without cache, but if cron runs while you are disabling the module then php bin/magento commands don't work. ...and it is impossible to then run `php bin/magento cache:enable`.

**Comment by azeemism** [comment]
@Ctucker9233 @peterhirn you need to stop cron before carrying out an update or you may run into install issues. Deleting `var/di` while important during an install procedure, does not solve the issue.

@peterhirn, you had to replace env.php as the first step before your troubleshooting was able to work around "consequent"/"or additional" issues. So your workaround produces a temporary fix not a solution.

**Comment by DanielRuf** [comment]
This still happens in some cases.
I guess this is some race condition and missing lockfile checks when an install / deployment and a cronjob access the setup at the same time.

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-tomash. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97551`, `MAGETWO-97552` were created

**Comment by DanielRuf** [comment]
If your env.php file is modified by any other process and has the `return 1;` line you can remove it as a workaround or recover an empty env file which has only this line like in https://github.com/magento/magento2/issues/7165#issuecomment-265120173

This is a bug due to race conditions and missing locks and too many processes which access and write to the env file.

**Comment by magento-engcom-team** [comment]
Hi @azeemism. 

Thank you for your report and collaboration!

The related internal Jira ticket `MAGETWO-97552` was closed as `non-reproducible` in `2.3-develop`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean Magento instance or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.

**Comment by tuyennn** [comment]
@DanielRuf have same problem
```
<?php
return 1;
```
What the hell here with Magento
